diff --git a/node_modules/express-validation/lib/index.js b/node_modules/express-validation/lib/index.js
index 50c3e02..5701f4a 100644
--- a/node_modules/express-validation/lib/index.js
+++ b/node_modules/express-validation/lib/index.js
@@ -36,7 +36,7 @@ exports.validate = (schema = {}, options = {}, joi = {}) => {
         }
       }
 
-      const errors = keyByField(clean(await Promise.all(promises)), options.keyByField);
+      const errors = keyByField(request, clean(await Promise.all(promises)), options.keyByField);
 
       if (errors) {
         next(new ValidationError(errors, evOptions));
diff --git a/node_modules/express-validation/lib/reducers.js b/node_modules/express-validation/lib/reducers.js
index 60b48af..c393679 100644
--- a/node_modules/express-validation/lib/reducers.js
+++ b/node_modules/express-validation/lib/reducers.js
@@ -3,11 +3,11 @@ exports.clean = (results) => results
   .reduce((a, b) => a.concat(b), [])
   .reduce((acc, curr) => ({ ...acc, ...curr }), null);
 
-exports.keyByField = (results, reduce) => {
+exports.keyByField = (request, results, reduce) => {
   if (results && reduce) {
     return Object.keys(results)
       .map(parameter => results[parameter]
-        .reduce((a, b) => Object.assign(a, { [b.context.key]: b.message }), {}));
+        .reduce((a, b) => Object.assign(a, { [b.context.key]: request.t(b.message) }), {}));
   }
 
   return results;
