// @jorgebodega/typeorm-seeding/dist/factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Factory = void 0;
const datasource_1 = require("./datasource");
const instanceAttribute_1 = require("./instanceAttribute");
const lazyInstanceAttribute_1 = require("./lazyInstanceAttribute");
const subfactory_1 = require("./subfactory");
/**
 * @deprecated Starting from v6, this class is moved to [typeorm-factory](https://github.com/jorgebodega/typeorm-factory)
 */
class Factory {
    /**
     * Make a new entity without persisting it
     */
    async make(overrideParams = {}) {
        const attrs = { ...this.attrs(), ...overrideParams };
        const entity = await this.makeEntity(attrs, false);
        await this.applyLazyAttributes(entity, attrs, false);
        return entity;
    }
    /**
     * Make many new entities without persisting it
     */
    async makeMany(amount, overrideParams = {}) {
        const list = [];
        for (let index = 0; index < amount; index++) {
            list[index] = await this.make(overrideParams);
        }
        return list;
    }
    /**
     * Create a new entity and persist it
     */
    async create(overrideParams = {}, saveOptions) {
        const attrs = { ...this.attrs(), ...overrideParams };
        const preloadedAttrs = Object.entries(attrs).filter(([, value]) => !(value instanceof lazyInstanceAttribute_1.LazyInstanceAttribute));
        const entity = await this.makeEntity(Object.fromEntries(preloadedAttrs), true);
        const em = (0, datasource_1.fetchDataSource)().createEntityManager();
        const savedEntity = await em.save(entity, saveOptions);
        await this.applyLazyAttributes(savedEntity, attrs, true);
        return em.save(savedEntity, saveOptions);
    }
    /**
     * Create many new entities and persist them
     */
    async createMany(amount, overrideParams = {}, saveOptions) {
        const list = [];
        for (let index = 0; index < amount; index++) {
            list[index] = await this.create(overrideParams, saveOptions);
        }
        return list;
    }
    async makeEntity(attrs, shouldPersist) {
        const entity = new this.entity();
        await Promise.all(Object.entries(attrs).map(async ([key, value]) => {
            Object.assign(entity, { [key]: await Factory.resolveValue(value, shouldPersist) });
        }));
        await Promise.all(Object.entries(attrs).map(async ([key, value]) => {
            if (value instanceof instanceAttribute_1.InstanceAttribute) {
                const newAttrib = value.resolve(entity);
                Object.assign(entity, { [key]: await Factory.resolveValue(newAttrib, shouldPersist) });
            }
        }));
        return entity;
    }
    async applyLazyAttributes(entity, attrs, shouldPersist) {
        await Promise.all(Object.entries(attrs).map(async ([key, value]) => {
            if (value instanceof lazyInstanceAttribute_1.LazyInstanceAttribute) {
                const newAttrib = value.resolve(entity);
                Object.assign(entity, { [key]: await Factory.resolveValue(newAttrib, shouldPersist) });
            }
        }));
    }
    static resolveValue(value, shouldPersist) {
        if (value instanceof subfactory_1.Subfactory) {
            return shouldPersist ? value.create() : value.make();
        }
        else if (value instanceof Function) {
            return value();
        }
        else {
            return value;
        }
    }
}
exports.Factory = Factory;
